hphp/runtime/vm/jit/vasm-arm.cpp

```diff
@@ -620,46 +655,54 @@ void Vgen::emit(const cloadq& i) {
  * In the following implementation,
  *   N, Z, V are updated according to result
  *   C is cleared (FIXME)
- *   PF, AF are not available
  */
 void Vgen::emit(const imul& i) {
 
   // Do the multiplication
   a->Mul(X(i.d), X(i.s0), X(i.s1));
 
-  vixl::Label after;
-  vixl::Label checkSign;
-  vixl::Label Overflow;
-
-  // Do the multiplication for the upper 64 bits of a 128 bit result.
-  // If the result is not all zeroes or all ones, then we have overflow.
-  // If the result is all zeroes or all ones, and the sign is the same,
-  // for both hi and low, then there is no overflow.
-  a->smulh(rAsm, X(i.s0), X(i.s1));
-
-  // If hi is all 0's or 1's, then check the sign, else overflow (fallthrough)
-  a->Cmp(rAsm, 0);
-  a->B(&checkSign, vixl::eq);
-  a->Cmp(rAsm, -1);
-  a->B(&checkSign, vixl::eq);
-
-  // Overflow, so conditionally set N and Z bits and then or in V bit.
-  a->Bind(&Overflow);
-  a->Bic(vixl::xzr, X(i.d), vixl::xzr, SetFlags);
-  a->Mrs(rAsm, NZCV);
-  a->Orr(rAsm, rAsm, 1<<28);
-  a->Msr(NZCV, rAsm);
-  a->B(&after);
-
-  // Check the signs of hi and lo.
-  a->Bind(&checkSign);
-  a->Eor(rAsm, rAsm, X(i.d));
-  a->Tbnz(rAsm, 63, &Overflow);
+  // If we have to set any flags, then always set N and Z since it's cheap.
+  // Only set V when absolutely necessary. C is not supported.
+  if (i.fl) {
```

Comment:
Avoids all synthetic status flag setting if there are no SF bits required. If even one is required, then the relatively cheap Z and N are checked.